src/garage/torch/distributions/tanh_normal.py

```diff
@@ -0,0 +1,154 @@
+"""A Gaussian distribution with tanh transformation."""
+from collections import namedtuple
+
+import torch
+from torch.distributions import Distribution, MultivariateNormal
+
+
+class TanhNormal(Distribution):
+    # pylint: disable=abstract-method
+    """A gaussian distribution that has a tanh transformation applied to it.
+
+    Algorithms like SAC and Pearl use this transformed distribution.
+    It can be thought of as a distribution of X where
+        Y ~ N(mean, cov)
+        X ~ tanh(Y)
+
+    Args:
+        mean(torch.tensor): The mean of this distribution.
+        cov(torch.tensor): The covariance of this distribution.
+        epsilon(float): Coefficient of stability used when calculating
+                        the log likehilhood of a sample on a distribution.
+
+    """
+
+    def __init__(self, mean, cov, epsilon=1e-6):
+        self._normal = MultivariateNormal(mean, cov)
+        self._epsilon = epsilon
+        super().__init__()
+
+    def log_prob(self, value, pre_tanh_value=None):
+        # pylint: disable=arguments-differ
+        """The log likelihood of a sample on the this Tanh Distribution.
+
+        Args:
+            value(torch.tensor): The sample whose loglikelihood is being
+                                 computed.
+            pre_tanh_value(torch.tensor): The value prior to having the tanh
+                                          function applied to it but after it
+                                          has been sampled from the normal
+                                          distribution.
+
+        Notes:
+            - pre_tanh_value should usually not be None, but as the option
+            available just in case a value is not/cannot be passed.
+
+        Returns:
+            torch.tensor: log likelihood.
+
+        """
+
+        def _clip_but_pass_gradient(x, lower=0., upper=1.):
+            """Clipping function that allows for gradients to flow through.
+
+            Args:
+                x(torch.tensor): value to be clipped
+                lower(float): lower bound of clipping
+                upper(float): upper bound of clipping
+
+            Returns:
+                torch.tensor: x clipped between lower and upper.
+
+            """
+            clip_up = (x > upper).float()
+            clip_low = (x < lower).float()
+            with torch.no_grad():
+                clip = ((upper - x) * clip_up + (lower - x) * clip_low)
+            return x + clip
+
+        if pre_tanh_value is None:
+            pre_tanh_value = torch.log((1 + value) / (1 - value)) / 2
+        ret = (
+            self._normal.log_prob(pre_tanh_value) - torch.sum(
+                torch.log(
+                    _clip_but_pass_gradient((1. - value**2))  # noqa: E128 E501
+                    + self._epsilon),
+                axis=-1))
+        return ret
+
+    def sample(self, return_pre_tanh_value=False):
```

Comment:
to keep parity with the `Distribution` API, you can implement `sample` without the new argument/return value and then add some other method e.g. `sample_and_transform` with the semantics you want.